Sources/Fullscreen/FullscreenGallery/FullscreenImageView.swift

```diff
@@ -0,0 +1,162 @@
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        setup()
+    }
+
+    convenience init() {
+        self.init(frame: .zero)
+    }
+
+    private func setup() {
+        delegate = self
+        showsHorizontalScrollIndicator = false
+        showsVerticalScrollIndicator = false
+
+        addGestureRecognizer(doubleTapRecognizer)
+
+        addSubview(imageView)
+        imageView.fillInSuperview()
+    }
+
+    // MARK: - Public methods
+
+    public func superviewWillTransition(to size: CGSize) {
+        recalculateLimitsAndBounds(forSize: size)
+    }
+
+    public func recalculateLimitsAndBounds() {
+        recalculateLimitsAndBounds(forSize: bounds.size)
+    }
+
+    // MARK: - Private methods
+
+    private func recalculateLimitsAndBounds(forSize size: CGSize) {
+        calculateZoomLimits(forViewSize: size)
+        adjustImageInsets(forViewSize: size)
+    }
+
+    private func calculateZoomLimits(forViewSize viewSize: CGSize) {
+        guard let image = self.image else {
+            return
+        }
+
+        let constrainedImageSize: CGFloat
+        let constrainedScreenSize: CGFloat
+
+        let screenAspectRatio = viewSize.width / viewSize.height
+        let imageAspectRatio = image.size.width / image.size.height
+
+        if screenAspectRatio < imageAspectRatio {
+            constrainedImageSize = image.size.width
+            constrainedScreenSize = viewSize.width
+        } else {
+            constrainedImageSize = image.size.height
+            constrainedScreenSize = viewSize.height
+        }
+
+        let minZoomScale = constrainedScreenSize / constrainedImageSize
+        let maxZoomScale = max(2.5 * minZoomScale, (2.0 * constrainedImageSize) / constrainedScreenSize)
+
+        maximumZoomScale = maxZoomScale
+        minimumZoomScale = minZoomScale
+        zoomScale = minZoomScale
+
+        // The contentSize will be set automatically at a later point in time, but in order
+        // to be able to work with it immediately we need to set it manually.
+        contentSize = CGSize(width: minZoomScale * image.size.width, height: minZoomScale * image.size.height)
+    }
+
+    private func adjustImageInsets(forViewSize viewSize: CGSize) {
+        let offsetX = max((viewSize.width - contentSize.width) * 0.5, 0)
+        let offsetY = max((viewSize.height - contentSize.height) * 0.5, 0)
+        contentInset = UIEdgeInsets(top: offsetY, leading: offsetX, bottom: 0, trailing: 0)
+    }
+}
+
+// MARK: - Double tap handling
+
+extension FullscreenImageView {
+    @objc private func onDoubleTap(_ recognizer: UIGestureRecognizer) {
+        // When double tapping, only alternate between 1x and 2x. Further zooming may be permitted,
+        // but that has to be done explicitly. Double tapping an already zoomed in image should always
+        // result in zooming back out to 1x.
+        var newZoom: CGFloat
+        if zoomScale >= minimumZoomScale * 1.05 {
```

Comment:
What does `1.05` mean? Could this be placed into a variable, perhaps?

Reply:
This is a check to see if the view is zoomed in more than 5% – not sure if this warrants its own constant tbh.

Reply:
The 5% is also completely arbitrary – what I'm really wanting to do is get rid of floating-point errors, so the check could _(should?)_ **really** have been:

`if zoomScale >= minimumZoomScale * 1.001 {`

Reply:
If not a variable, a comment should be enough. Just to explain to people in the future why you're multiplying `minimumZoomScale` by an arbitrary value😅

Reply:
Floating point errors like value being 0.9999997 (or something) instead of 1? You could try .round() or .rounded() then

Reply:
`zoomScale` and `minimumZoomScale` will hold arbitrary float values though _(often in the range `0.3...1.0`)_, so rounding them would remove all information from them.

Reply:
Played around with this a bit more – if the user has accidentally zoomed in slightly _(2-3%)_, it feels "off" that double-tapping zooms out. I'm changing the limit to 10% – if you're zoomed in less than this limit a double-tap will zoom **in**.